Added warnings handling

rahim-kanji opened the pull request:
Resolves [Issue#1751](https://github.com/sysown/proxysql/issues/1751)

### Summary:
When a query is executed, and it generates a warning, the multiplexing feature is temporarily disabled. The status is set to `'has_warnings'` and the `'warning_in_hg'` is assigned the current host group. If the next query happens to be either `'SHOW WARNINGS'` or `'SHOW COUNT(*) FROM WARNINGS'`, it will be executed on the host group specified in `'warning_in_hg'`. However, if any other type of query is executed, multiplexing is re-enabled, and `'warning_in_hg'` is reset to -1.

### Changes:
* Implemented `'SHOW WARNINGS'` and `'SHOW COUNT(*) FROM WARNINGS'` query handling.
* When a query produces warnings, the warning count is now included in the query response.

For more granular control, a new variable `'mysql-handle_warnings' `and `'handling_warnings'` flag in the `'mysql_hostgroup_attributes'` table (`'hostgroup_settings'`) has been introduced to control warnings handling in ProxySQL: 
- A value of `'0'` means warnings handling is completely disabled.
- A value of `'1'` enables warnings handling as described above.
> _`'handling_warnings'` flag holds top priority and can override the value of the global variable `'mysql_handle_warnings'`_

Additionally, a new variable `'mysql-query_cache_handle_warnings'` has been introduced for the query cache feature:
  - A value of `'0'` means that if a query execution produces warnings, the resultset will not be saved in the query cache.
  - A value of `'1'` indicates that if a query produces warnings, the resultset will be saved in the query cache, but the warning count in that result set will be set to `'0'`.

### Note:
* Warning handling is dependent on the query digest and `'mysql-query_digests'` must be enabled for it to work.
* Warnings saved in logfile [`'mysql-log_mysql_warnings_enabled'`] and the current feature can work in conjunction with each other.
* This feature will not work correctly if  `'mysql-query_digests_keep_comment'` is enabled, and `'SHOW WARNINGS'` or `'SHOW COUNT(*) FROM WARNINGS'` is used with comments.
* No support for handling `'SELECT @@warning_count'`. If this query is executed, multiplexing will be disabled and will remain disabled for that connection.
* With the exception of a few, all TAP tests are now geared up to log warning count and the query that triggered the warning during its execution.

rahim-kanji committed:
Enhanced Warnings handling in ProxySQL:

* Implemented 'SHOW WARNINGS' and 'SHOW COUNT(*) FROM WARNINGS' query handling.
* When a query produces warnings, the warning count is now included in the query response.

Additionally, a new variable 'mysql-query_cache_with_warnings_support' has been introduced for the query cache feature:

* A value of '0' means that if a query execution produces warnings, the resultset will not be saved in the query cache.
* A value of '1' indicates that if a query produces warnings, the resultset will be saved in the query cache, but the warning count in that resultset will be set to '0'.

** Note:
* Warning handling is dependent on the query digest and 'mysql-query_digests' must be enabled for it to work.
* Warnings saved in logfile ['mysql-log_mysql_warnings_enabled'] and the current feature can work in conjunction with each other.

rahim-kanji committed:
Added TAP test

rahim-kanji committed:
Fixed code style

rahim-kanji committed:
Added delay between query execution

rahim-kanji committed:
Changed digest to match_digest

rahim-kanji committed:
Remove *.so files from tap folder on make clean

rahim-kanji committed:
Remove old mysql query rule entries

renecannao commented on lib/MySQL_Session.cpp:
```diff
@@ -1538,19 +1542,41 @@ bool MySQL_Session::handler_special_queries(PtrSize_t *pkt) {
-		resultset->add_column_definition(SQLITE_TEXT,"Message");
+	// if query digest is disabled, warnings in ProxySQL are also deactivated, 
+	// resulting in an empty response being sent to the client.
+	if ((pkt->size == 18) && (strncasecmp((char*)"SHOW WARNINGS", (char*)pkt->ptr + 5, 13) == 0) &&
```
Since we check `CurrentQuery.QueryParserArgs.digest_text` , maybe it is the right opportunity to perform a string compare on the `digest_text` and not on the packet itself . This will allows us to match also in case of comments.
Also, maybe it makes sense to move all this code in `handler___status_WAITING_CLIENT_DATA___STATE_SLEEP___MYSQL_COM_QUERY_qpo()`

rahim-kanji replied:
This part will only be executed if digest_text is null ('mysql-query_digests'  is disabled)

renecannao commented on lib/MySQL_Session.cpp:
```diff
@@ -1538,19 +1542,41 @@ bool MySQL_Session::handler_special_queries(PtrSize_t *pkt) {
+	}
+	// if query digest is disabled, warnings in ProxySQL are also deactivated, 
+	// resulting in zero warning count sent to the client.
+	if ((pkt->size == 27) && (strncasecmp((char*)"SHOW COUNT(*) WARNINGS", (char*)pkt->ptr + 5, 22) == 0) &&
```
Same as previous comment

rahim-kanji replied:
This part will only be executed if digest_text is null ('mysql-query_digests'  is disabled)

renecannao commented on lib/MySQL_Session.cpp:
```diff
@@ -5985,26 +6011,82 @@ bool MySQL_Session::handler___status_WAITING_CLIENT_DATA___STATE_SLEEP___MYSQL_C
+			proxy_debug(PROXY_DEBUG_MYSQL_COM, 5, "Intercepted '%s'\n", dig);
+			if (warning_in_hg > -1) {
+				proxy_debug(PROXY_DEBUG_MYSQL_COM, 5, "Changing current_hostgroup to '%d'\n", warning_in_hg);
+				current_hostgroup = warning_in_hg;
```
I think we can reply without forwarding the query to the backend

rahim-kanji replied:
Done

renecannao commented on lib/mysql_connection.cpp:
```diff
@@ -2569,6 +2572,34 @@ void MySQL_Connection::ProcessQueryAndSetStatusFlags(char *query_digest_text) {
+	// checking warnings and disabling multiplexing will be effective only when the mysql-query_digests is enabled
+	if (myds->sess->CurrentQuery.QueryParserArgs.digest_text) {
+		if (get_status(STATUS_MYSQL_CONNECTION_HAS_WARNINGS) == false) {
+			if (mysql_warning_count(this->mysql) > 0) {
```
Because calling `mysql_warning_count()` we need to ensure that `this->mysql` is not `NULL` .

rahim-kanji replied:
Done

renecannao commented on test/tap/tests/test_warnings-t.cpp:
```diff
@@ -0,0 +1,347 @@
+		} \
+	} while(0)
+
+#define MYSQL_CLEAR_RESULT(mysql) do { MYSQL_RES* mysql_result = mysql_use_result(mysql); \
```
Instead of `mysql_use_result()` , then looping through all the rows with `mysql_fetch_row()` , you can simply use `mysql_store_result()`:
https://dev.mysql.com/doc/c-api/8.0/en/mysql-store-result.html

At the end, use `mysql_free_result()` anyway.

rahim-kanji replied:
Done

renecannao commented on test/tap/tests/test_warnings-t.cpp:
```diff
@@ -0,0 +1,347 @@
+		return -1;
+	}
+
+	plan(9+13+4+23+9+10);
```
Please add some explanation to these numbers.

rahim-kanji replied:
Done

rahim-kanji committed:
Few fixes

rahim-kanji committed:
Merge branch 'v2.x_hostgroup_attr_settings' into v2.x_warnings_support

rahim-kanji committed:
* Updated variable name from 'mysql-query_cache_with_warnings_support' to 'mysql-query_cache_handle_warnings'
* Introduced global variable 'mysql-handle_warnings' to control warning support in ProxySQL
* Included 'handling_warnings' flag in the 'mysql_hostgroup_attributes' table (hostgroup_settings). This flag holds top priority and can override the value of the global variable 'mysql_handle_warnings'.

rahim-kanji committed:
Updated cluster TAP test

rahim-kanji committed:
Merge branch 'v2.x_hostgroup_attr_settings' into v2.x_warnings_support

rahim-kanji committed:
Updated test_mysql_hostgroup_attributes-1-t TAP test

rahim-kanji committed:
* Update warning_count as soon as query is executed.

rahim-kanji committed:
* Setting 'warning_count' as soon as query is executed.
* 'MySQL_Connection::warning_count' will be globally employed for managing multiplexing control and transmitting the warning count to the client
* Fixed prepare statement bug: When preparing, if the digest matches, the result set is retrieved from the cache and sent to the client.

Note:
Currently using 'mysql_warning_count' to retrieve the warning count for prepared statement.

rahim-kanji committed:
Merge branch 'v2.x_hostgroup_attr_settings' into v2.x_warnings_support

kmcphillips commented:
I'm not in a position to review this, but just wanted to say we tripped over this issue recently and this fix is exactly what we are looking for. So thank you. 🙏

rahim-kanji committed:
* Execute all test cases with 'CLIENT_DEPRECATE_EOF' both enabled and disabled.
* Introduced 'ps_type' enum to differentiate between preparing a statement and executing a statement
* Fixed issue where the warning count fails to reset when the SET statement generates a warning.
* Resolved the issue where the 'server_status' and 'warning count' were not accurately represented in the statement.
* Replaced 'mysql_warning_count' with 'mysql_stmt_warning_count' for retrieving the warning count in statements.
* Added 'handle_warning' field in hostgroup_attributes TAP test
* Updated warnings TAP test

mirostauder commented:
retest this please

mirostauder commented:
retest this please

rahim-kanji commented:
retest this please

rahim-kanji commented:
retest this please

rahim-kanji commented:
retest this please

rahim-kanji committed:
* If the backend server has CLIENT_DEPRECATE_EOF enabled, and the client does not support CLIENT_DEPRECATE_EOF, warning_count will be '0' in intermediate EOF packet.
* Suppress the display of warning counts in the "SHOW WARNINGS" statement.

rahim-kanji committed:
Fixed test_wexecvp_syscall_failures-t compiling

mirostauder commented:
retest this please

mirostauder commented:
retest this please

rahim-kanji committed:
* Overriding some of the mariadb APIs to extract the warning count and print it in the log. This override will apply to all TAP tests, except when the TAP test is linked with the MySQL client library (LIBMYSQL_HELPER defined).
* Avoid closing of same mysql connection twice on error.
* Disabled warning count logging on few TAP tests
* Enable CLIENT_DEPRECATE_EOF on 20% of total mysql connections

rahim-kanji committed:
Merge remote-tracking branch 'Master/v2.x' into v2.x_warning_support_final

# Conflicts:
#	test/tap/tests/set_testing-240-t.cpp

rahim-kanji committed:
Added README

mirostauder commented:
retest this please

rahim-kanji commented:
retest this please

mirostauder commented:
retest this please
